### ISSUE TITLE
[API Proposal]: Add `.All` and `.Any()` for bool collections in System.Linq.Enumerable

### ISSUE THREAD

**Issue by steveberdy** [dotnet/runtime#69925]
### Background and motivation

I noticed that to use `Any` or `All` with a collection of booleans you still have to provide a predicate.
If you wanted to check for any `true` values in a collection called `boolArray`, you would have to write `boolArray.Any(b => b)`.
It would be nice to add overloaded extension methods `Any()` and `All()` for just `IEnumerable<bool>` types.

### API Proposal

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Linq/src/System/Linq/AnyAll.cs
```csharp
namespace System.Linq
{
    public static partial class Enumerable
    {
        public static bool Any(this IEnumerable<bool> source) { }
        public static bool All(this IEnumerable<bool> source) { }
    }
}
```

### API Usage

var array = new bool[] { true, false, true };
Console.WriteLine(array.Any());
Console.WriteLine(array.All());

**Comment by huoyaoyuan** [comment]
We can widen it to cover any values, instead of predicates:

```csharp
namespace System.Linq
{
    public static partial class Enumerable
    {
        public static T Any<T>(this IEnumerable<T> source, T value) { }
        public static T All<T>(this IEnumerable<T> source) { }
    }
}
```

Then we can say:

```csharp
array.All(true);
array.Any(false);
```

The functions overlaps for bool, but can be useful for other types.

**Comment by tfenise** [comment]
There is already `public static bool Any<TSource>(this IEnumerable<TSource> source)`. This would be a breaking change.

**Comment by steveberdy** [comment]
Yes, that's true, there would be a conflict there. I forgot about that overload even though I've seen it in source and have used it before.

@huoyaoyuan those API additions would be great. I think the signatures are off though:
```csharp
public static bool Any<T>(this IEnumerable<T> source, T value) { }
public static bool All<T>(this IEnumerable<T> source, T value) { }
```

Where would the overlap be for bool?

**Comment by steveberdy** [comment]
If only this...
```csharp
public static bool Any<TSource>(this IEnumerable<TSource> source)
```
Were this...
```csharp
public static bool IsEmpty<TSource>(this IEnumerable<TSource> source)
```
Then we could have...
```csharp
public static bool Any(this IEnumerable<bool> source)
```

Too late though haha.

**Comment by stephentoub** [comment]
> `public static bool Any<TSource>(this IEnumerable<TSource> source, TSource value) { }`

How is that different from Enumerable.Contains?

**Comment by steveberdy** [comment]
> How is that different from Enumerable.Contains?

It's not, thanks for pointing that out. It would still be nice to have
```csharp
public static bool All<TSource>(this IEnumerable<TSource> source, TSource value) { }
```

**Comment by stephentoub** [comment]
> It would still be nice to have

What's the use case?  If it's the bool example that started this issue, isn't the proposed `All(true)` just `!Contains(false)`?

**Comment by steveberdy** [comment]
I suppose the use case isn't all that clear besides making LINQ's `Any` and `All` more like python's `any` and `all`. After all, there is almost never one way to do something with LINQ.

**Comment by eiriktsarpalis** [comment]
I don't believe the proposed methods are common enough to warrant inclusion in System.Linq. We came to a [similar conclusion](https://github.com/dotnet/runtime/issues/54220#issuecomment-968834849) when debating `.Flatten()`, since it also can be expressed trivially using `.SelectMany(x => x)`.

Related to https://github.com/dotnet/runtime/issues/67194
